File: merge_sort/python/merge_sort.py
import math

def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    
    middle = math.floor(len(arr) / 2)
    left = arr[:middle]
    right = arr[middle:]

    return merge(merge_sort(left), merge_sort(right))


def merge(left, right):
    results = []
    indexLeft = 0
    indexRight = 0

    while len(left) > indexLeft and len(right) > indexRight:
        if left[indexLeft] < right[indexRight]:
            results.append(left[indexLeft])
            indexLeft += 1
        else:
            results.append(right[indexRight])
            indexRight += 1
    
    results += left[indexLeft:] + right[indexRight:]

    return results


numArr = [38, 27, 43, 39, 82, 10, 5]

print(merge_sort(numArr))

File: merge_sort/javascript/merge_sort.js
const mergeSort = (arr) => {
  if(arr.length <= 1) {
    return arr;
  }
  
  let middle = Math.floor(arr.length / 2);
  let left = arr.slice(0, middle);
  let right = arr.slice(middle, arr.length);
  
  return merge(mergeSort(left), mergeSort(right));
}

const merge = (left, right) => {
  let results = [];
  let indexLeft = 0;
  let indexRight = 0;
  
  while(left.length > indexLeft && right.length > indexRight) {
    if(left[indexLeft] < right[indexRight]) {
      results.push(left[indexLeft]);
      indexLeft++;
    } else {
      results.push(right[indexRight]);
      indexRight++;
    }
  }
  
  return results.concat(left.slice(indexLeft)).concat(right.slice(indexRight));
}

const numArr = [38, 27, 43, 39, 82, 10, 5]

console.log(mergeSort(numArr))